### PR TITLE
query: add initial time-based query pushdown

### PR DESCRIPTION
**Pull request by GiedriusS**
Add time-based pushdown mechanism. It works by directly querying a
Thanos Store node if it is the only one partially or fully matching the
time range of a query. I believe that this will be the case most of the
times because typically Sidecar/Ruler/Receive cover a few days of time
range and horizontally balanced Thanos Store instances cover the
historical data from remote object storage. With query-frontend in front
splitting the queries into smaller parts means that with higher time ranges, most of
the queries can be simply pushed down. For example, with Sidecars
covering 2 days of data, with a 7d query, and a split interval of 1d
this means that 5/7 * 100 = ~71% of processing can be pushed down to the
leaf nodes.

Small ad-hoc tests which touch a lot of timeseries show a 2x reduction
in query duration.

Hide this under an "experimental feature" flag
because this still doesn't have everything implemented intentionally to have a smaller change:

* `lazySeriesSet` for concurrent `Select()`s in Thanos Store
* Missing some of the Prometheus engine flags
* Support the selection of StoreAPI nodes in the UI (use only them, not
all endpoints)
* No chunks/series limiters

Also, this API could be potentially implemented in other components.

If the approach looks good then we can merge this and implement these in follow-up PRs.

The main parts of this change is in:

* New QueryAPI service
* New `pkg/pushdown/`
* API request handling code

### CODE REVIEW

**Review by yeya24 — COMMENTED** [review]
Love this work! Can you add another E2E test case between query and store gateway with and without query pushdown? Is the returned result the same for range query?

**Comment by yeya24** [comment]
We are filtering stores using the given time range in the API. But when promql engine performs the Select call, will it use the same time range given from the API?

https://github.com/prometheus/prometheus/blob/f103acd5135b8bbe885b17a73dafc7bbb586319c/promql/engine.go#L762
From this function, IIUC the time range will be adjusted based on the promql queries. WDYT?

**Comment by GiedriusS** [comment]
> We are filtering stores using the given time range in the API. But when promql engine performs the Select call, will it use the same time range given from the API?
> 
> https://github.com/prometheus/prometheus/blob/f103acd5135b8bbe885b17a73dafc7bbb586319c/promql/engine.go#L762 From this function, IIUC the time range will be adjusted based on the promql query itself. WDYT?
> 
> If we can extract this logic from promql, then maybe it is possible to pushdown instant queries as well.

That's true but I'd like to start simple and cover only one case at the beginning as in this PR. Maybe you could add this idea to https://github.com/thanos-io/thanos/issues/305?

File: pkg/query/endpointset.go [hunk 1]
```diff
@@ -146,6 +147,7 @@ func (es *grpcEndpointSpec) fillExpectedAPIs(componentType component.Component,
 				MinTime: mintime,
 				MaxTime: maxTime,
 			},
+			Query: &infopb.QueryInfo{},
```
Comment (GiedriusS) [inline pkg/query/endpointset.go:150]: Note that without `--endpoint` it means that we assume that all Stores implement QueryAPI :s

File: pkg/pushdown/querypb/rpc.proto [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.goproto_getters_all) = false;
+// Do not generate XXX fields to reduce memory footprint and opening a door
+// for zero-copy casts to/from prometheus data types.
+option (gogoproto.goproto_unkeyed_all) = false;
+option (gogoproto.goproto_unrecognized_all) = false;
+option (gogoproto.goproto_sizecache_all) = false;
+
+service Query {
```
Comment (bwplotka) [inline pkg/pushdown/querypb/rpc.proto:19]: OK, I think the main question we have to ask ourselves - do we need another API, let me explain on main comment.

**Review by bwplotka — COMMENTED** [review]
OK, I think the main question we have to ask ourselves - do we need another API.

The thing is that StoreAPI is well equipped for the work - the pipeline is ready for doing global eval, local eval or mixed. Let's go through different executions for query `count(metric1{target=ext1})`. This makes PromQL engine selects in Querier  `metric1{target=ext1}` (and with `count` function in hints). Then Querier proxy code (StoreAPI querier implementation) fanouts to all matching stores. And let's say for simple case. There is only ONE matching store. And no duplicated metrics. We can then:

1.  Ask this matching store for non-pushdown StoreAPI call.  This will return all selected `metric1` series from remote store. We pass that as it is to PromQL engine in StoreSet. Then PromQL counts them up and returns result.

2. But in the same time, we can do a special StoreAPI.Series call  with e.g pushdown=true field that will tell the underlying store  to give us count of all metric1 series. So it will return us single series 
 for `count(metric1{target=ext1})` as long as we know there are no duplicates across fanouts, we can put that as a result to PromQL engine. PromQL engine will take that as only series count it (but it's only one, so fine) and return to response.

We can enhance 2 and even pass step so only necessary samples are returned that would be use by PromQL. 

I know it might be complex, so maybe a second method to service makes sense, but I would really really like to avoid another API - too much complexity IMO 🤔

**Comment by bwplotka** [comment]
If we would just  expose QuerAPI through gRPC - this might work, but only for super limited cases, where is only one Store matched. We block (or need separate path) for cases where we have `count, sum` etc function and multiple stores.. I think this is bit of missed opportunity - and we sacrifice a huge complexity if we would have now: `QueryAPI, StoreAPI and StorePushdownAPI`... etc

**Comment by yeya24** [comment]
> But in the same time, we can do a special StoreAPI.Series call with e.g pushdown=true field that will tell the underlying store to give us count of all metric1 series. So it will return us single series
for count(metric1{target=ext1}) as long as we know there are no duplicates across fanouts, we can put that as a result to PromQL engine. PromQL engine will take that as only series count it (but it's only one, so fine) and return to response.

PromQL engine only gets series set as result and it will try to apply `count` again. How to bypass this and tell it the result was already calculated? I feel if we don't have another API other than `Select` we cannot achieve query pushdown.

**Comment by bwplotka** [comment]
What else do you want than `series` result as input to PromQL? (: If global PromQL can take raw series but also pre-calculated series from another PromQL engine, in same way, no? (:

**Comment by GiedriusS** [comment]
> OK, I think the main question we have to ask ourselves - do we need another API.
> 
> ...
> I know it might be complex, so maybe a second method to service makes sense, but I would really really like to avoid another API - too much complexity IMO thinking

I understand you and remember this idea from your design document but are these two things exclusive to each other? I really think that we can have both. Just like most of the databases do optimizations at many layers, I think we should do both (and much more). The optimization in this PR is dead simple and quite trivial - the only complex part IMO is implementing the PromQL engine.

To be honest, there is probably no clear-cut answer. But this pull request would solve already many problems in at least two big deployments of Thanos that I know of. Implementing pushdown has been the longest time on our backlog but it has never been properly implemented - perhaps that is because the idea in its entirety is quite complex and we must divide it into smaller parts as in this PR?

Also, I understand that we should strive to implement as generic solutions as possible to ultimately reduce the complexity so that it would be easier to understand the code and maintain it but also on the other hand IMHO sometimes optimization come in smaller pieces, and two extra function calls in `queryRange()` doesn't seem like a huge overhead at all. 

@yeya24 what's your opinion on this topic?

**Comment by yeya24** [comment]
> Also, I understand that we should strive to implement as generic solutions as possible to ultimately reduce the complexity so that it would be easier to understand the code and maintain it but also on the other hand IMHO sometimes optimization come in smaller pieces, and two extra function calls in queryRange() doesn't seem like a huge overhead at all.
@yeya24 what's your opinion on this topic?

They can coexist and I think it is okay to have a separate API for pushdown for now as it is under an experimental feature flag. In the short term, it is hard to hack the promql engine to achieve pushdown directly.

What we need to do is to ensure the results consistency between pushdown and normal query. Continuous benchmarking or tests would help so that we can compare the performance.

**Comment by yeya24** [comment]
> Both are just series with some samples, nothing else.

Right, I understand the final results are just some samples. But the promql engine will apply the promql functions, aggregators again even if some series are pre-aggregated. 
I think that's why we need a new API to bypass the series post-processing from the engine. If it is possible to have a hook at the engine side then that would be great.

**Comment by GiedriusS** [comment]
Could we split this PR then into smaller ones? I could start with a pull request that adds the query engine to Thanos Store with tests. Maybe that would be easier?

**Comment by bwplotka** [comment]
Anyway, I think we are still not sure if we need a different API or not. That is what blocks us from developing on main IMO. I would keep it on the feature branch still. We can try different approaches and compare.  I would move this PR as it is into one Thanos pushdown branch (not your forks's PR), then I can try reusing Series and we can experiment. You see I would love to avoid extra API because of complexities like https://github.com/thanos-io/thanos/issues/4780

We would need to implement all twice and it's not trivial code to fan out all. The SIngle pipeline would be nice, if possible.

**Comment by GiedriusS** [comment]
> > But the promql engine will apply the promql functions, aggregators again even if some series are pre-aggregated.
> 
> That is totally fine, no? We need to do that in order to aggregate pre-aggregated results (:

It's not fine to do that with functions. For example, if you do `sum_over_time`, where step < range, and then apply `sum_over_time` on top then the result will be the double of `sum_over_time`. :P

**Pull request closed without merging**